simulator-core/src/main/java/com/evbox/everon/ocpp/simulator/station/StationState.java

```diff
@@ -160,4 +163,23 @@ public String toString() {
                         .filter(connector -> connector.getId().equals(connectorId))
                         .findAny());
     }
+
+    /**
+     * Returns defensive copy of station state view
+     * @return copy of station state view
+     */
+    public StationState getView() {
+        return copyOf(stationStateView);
+    }
+
+    public void refreshView() {
+        this.stationStateView = copyOf(this);
```

Comment:
I'm not sure why do we need this copy here...

Reply:
To protect it from modification by readers

Reply:
Again this raises questions, Dragan had also doubts about this code. It would be more elegant to make the class thread-safe.

Reply:
Copying is done twice - in two methods. To protect it for reading it is sufficient to do the copy in getView method. Why do you need to do another copy in refreshView?

Reply:
you're right, it's not needed here

Reply:
getView can be called by different threads, refreshView is called only by station, once handle(request) is done

Reply:
What you are saying this code can not guarantee that. It has a data-race. To prevent this, locking has to be used. An example: CopyOnWriteArraList

Reply:
State is not thread-safe by design. It is encapsulated inside of station. Station is the facade for other threads and it provides access only to the copy of state

Reply:
okay, I checked your code. And this is what we have:
refreshView -- is called by consumer-thread
getView -- can be called by other threads (that hold station instance)

refreshView() makes a copy of this. Copying can be omitted here. 
getView() makes another copy - which is good.

However as I already mentioned the bug is to provide solution which brings thread-safety to the station state

Reply:
refreshView() needs to make a copy, this is a method that will be called by same thread that is writing to the state. If I do copy in getView() only - copying will be done by other threads, which is not thread-safe.

The bug was: 'Simulator: Station state is not thread-safe but accessed by multiple-threads'. In this PR it's been addressed.
Statement that stationState has to be thread-safe is not a bug, it's a new requirement or an improvement, which is arguable.

Reply:
what prevents you to do volatile-write without a copy? To be more precise why this code is not thread-safe: this.stationStateView = this;?

Reply:
it is an option, but do we want other threads to read transient state? I'd prefer not to

Reply:
How would other thread read that? volatile-write happends-before volatile-read.

Reply:
If I simply set it to reference current object - copying will be done only by thread-reader, at the the same moment when state is being modified. Refreshment of copy won't be needed in this case, but readers will see not only final state before/after message handling, but also transient (intermediate) state. There's no issue in general, but I find it more convenient if transition of state A->B->C->D is encapsulated by station and, let's say test, or any other reader can see only A-D

Reply:
Looks like I see what you mean.
The solution might show not actual data. But data of last copy. Basically your solution is racing against who reads or creates a copy first.